lib/SILOptimizer/Mandatory/Differentiation.cpp

```diff
@@ -382,6 +964,109 @@ class Differentiation : public SILModuleTransform {
 };
 } // end anonymous namespace
 
+/// Given a curry thunk application, clone the thunk to return a
+/// `@differentiable` function-typed value and apply the cloned thunk.
+///
+/// Curry thunk type: `(Self) -> (T, ...) -> U`.
+/// Cloned thunk type: `(Self) -> @differentiable (T, ...) -> U`.
+static SILValue promoteCurryThunkApplicationToDifferentiableFunction(
+    DifferentiationTransformer &dt, DifferentiableFunctionInst *dfi,
+    SILBuilder &builder, SILLocation loc, DifferentiationInvoker invoker) {
+  auto origFnOperand = dfi->getOriginalFunction();
+  auto parameterIndices = dfi->getParameterIndices();
+  auto &context = dt.getContext();
+  unsigned resultIndex = context.getResultIndex(dfi);
+
+  // Check for curry thunk application:
+  // - The original function operand must be an `apply` instruction.
+  // - The `apply` callee must be a `function_ref` instruction.
+  // - The callee must return a function-typed value.
+  auto *ai = dyn_cast<ApplyInst>(origFnOperand);
+  if (!ai)
+    return nullptr;
+  auto *thunkRef = dyn_cast<FunctionRefInst>(ai->getCallee());
+  if (!thunkRef)
+    return nullptr;
+  auto *thunk = thunkRef->getReferencedFunctionOrNull();
+  auto thunkTy = thunk->getLoweredFunctionType();
+  auto thunkResult = thunkTy->getSingleResult();
+  auto resultFnTy = thunkResult.getInterfaceType()->getAs<SILFunctionType>();
+  if (!resultFnTy)
+    return nullptr;
+
+  // Create a new curry thunk.
+  SILAutoDiffIndices desiredIndices(resultIndex, parameterIndices);
+  // TODO(TF-685): Use more principled mangling for thunks.
+  auto newThunkName = "AD__" + thunk->getName().str() +
+                      "__differentiable_curry_thunk_" + desiredIndices.mangle();
+
+  // Construct new curry thunk type with `@differentiable` function
+  // result.
+  auto diffResultFnTy = resultFnTy->getWithExtInfo(
+      resultFnTy->getExtInfo().withDifferentiabilityKind(
+          DifferentiabilityKind::Normal));
+  auto newThunkResult = thunkResult.getWithInterfaceType(diffResultFnTy);
```

Comment:
If there is a common pattern here, maybe put the construction of this function onto a helper on SILOptFunctionBuilder? The hope with that was to hide the low level SILFunction creation functions in favor of higher level routines so we don't have to have huge invocations of SILFunctionType::get.

Reply:
`SILOptFunctionBuilder` looks very minimal, so I'm not sure

Reply:
@dan-zheng I wrote that code and the reason why it is minimal now is that I was trying to create the scaffolding for what I just mentioned. Just no one has had the time to fix it up old uses. That being said, we should not make the problem worse.